Vostok.Logging.File/Configuration/SafeSettingsCache.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Vostok.Logging.File.Configuration
+{
+    internal class SafeSettingsCache
+    {
+        private readonly SafeSettingsProvider provider;
+        private readonly TimeSpan ttl = TimeSpan.FromSeconds(1);
+        private readonly bool enabled;
+        private volatile Lazy<FileLogSettings> container;
+
+        public SafeSettingsCache(Func<FileLogSettings> provider)
+        {
+            this.provider = new SafeSettingsProvider(provider);
+
+            enabled = this.provider.Get().EnableFileLogSettingsCache;
+
+            if (enabled)
+                ResetContainer();
+        }
+
+        public FileLogSettings Get() => enabled
+            ? container.Value
+            : provider.Get();
+
+        private void ResetContainer()
+        {
+            container = new Lazy<FileLogSettings>(
```

Comment:
The problem I see with this approach is that it will effectively stall many/all of the application threads every second. Every update is goin to cause logging threads to mutually block on a single lock which can be disastrous with high workloads if the settings provider delegate does not return immediately for some reason (GC, whatever). I reckon that all updates except for the very first one must happen asynchronously and not cause any app threads blocking.

Reply:
We could also try changing the `LazyThreadSafetyMode` to PublicationOnly. In this case LazyInitialization is lockfree, only one thread can initialize the value because of CAS operation at the end. Of course, it may cause multiple simultaneous calls to `SafeSettingsProvider`, but it might be better than just doing an asynchronous update every second regardless of the demand of FileLogSettings.

Reply:
> We could also try changing the LazyThreadSafetyMode to PublicationOnly.

Well, it would definitely by superior to current implementation.

> it might be better than just doing an asynchronous update every second regardless of the demand of FileLogSettings

It's not really hard to avoid unsolicited background updates, though. Imagine something along these lines:

```
private static readonly object CooldownToken = new object();

private volatile object updateCooldown;

...
...

// Somewhere in the primary method:

if (updateCooldown == null && Interlocked.CompareExchange(ref updateCooldown, CooldownToken, null))
{
    // launch async settings update, turn updateCooldown to null after a delay
}

```

Reply:
I'm afraid of one specific case: we did not use a log for a long time, and suddenly we want to write something.
In this scenario, we call primary method to get settings, launch an asynchronous update, but return old settings, because we want out method to be non-blocking.

It means that settings will be refreshed not only after 1 sec delay, but in fact after a delay and at least one log write, which is counter-intuitive and may be not really convinient to use.
 
However, we can probably expose something like `ForceRefresh` blocking method to avoid this problem.

Reply:
Yeah, I see what you're talking about. That is indeed a bit nasty. But I'm inclined to view this quirk as an acceptable tradeoff for guaranteed non-blocking behaviour. We should probably just document it and outline workarounds (disabling the cache or doing a manual forced refresh).